Add Number platform to Wallbox

hesselonline opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
In this change I add a number component to the Wallbox integration. After merging this PR the integration will have 2 platforms: number and sensor. (Long term goal is to also add a lock and switch platform). I also made a small overall improvement by making the WallboxHub class a child of DataUpdateCoordinator. Tests are included for the new platform.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/18446

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ X] The code change is tested and works locally.
- [ X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ X] There is no commented out code in this PR.
- [ X] I have followed the [development checklist][dev-checklist]
- [ X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

hesselonline committed:
Changed Testing, removed custom exception
Removed custom exceptions, reverted to builtin. Changed testing approach in all tests, now using the core interface to setup device and mock_requests to create test responses for all calls.

hesselonline committed:
Reintroduce InvalidAuth exception in __init__
Remove reference to internal HA exception, Reintroduce custom exception

hesselonline committed:
Change WallboxClass to coordinator

hesselonline committed:
Merge branch 'dev' of github.com:home-assistant/core into wallbox-change-coordinator

hesselonline committed:
Add number platform to Wallbox
Number platform was added in this commit, the WallboxHub is now a child class of DataUpdateCoordinator

hesselonline committed:
Add missing keys to manifest.json

hesselonline committed:
Fix pylance error, add public method to wallbox class.

hesselonline committed:
Added Reauth path in config_flow, number only available with admin user

hesselonline committed:
Add already configured to strings.json

hesselonline committed:
Delete .favorites.json

delete VSCode integration file

yayitazale commented:
27 days waiting with all the checks passed??? Please, some wallbox user need this PR to control their charger max amps.

yayitazale commented on homeassistant/components/wallbox/const.py:
```diff
@@ -1,14 +1,18 @@
+    DEVICE_CLASS_CURRENT,
+    DEVICE_CLASS_ENERGY,
+    DEVICE_CLASS_POWER,
     ELECTRICAL_CURRENT_AMPERE,
```
ELECTRIC_CURRENT_AMPERE

yayitazale commented on homeassistant/components/wallbox/const.py:
```diff
@@ -20,80 +24,87 @@
+        CONF_ICON: None,
         CONF_NAME: "Max Available Power",
         CONF_ROUND: 0,
         CONF_UNIT_OF_MEASUREMENT: ELECTRICAL_CURRENT_AMPERE,
```
ELECTRIC_CURRENT_AMPERE

hesselonline replied:
Thnx for the review, and, good catch. ELECTRICAL apparently has become ELECTRIC in the past weeks ;-)

yayitazale commented on homeassistant/components/wallbox/const.py:
```diff
@@ -20,80 +24,87 @@
+        CONF_ICON: None,
+        CONF_NAME: "Max. Charging Current",
+        CONF_ROUND: None,
+        CONF_UNIT_OF_MEASUREMENT: ELECTRICAL_CURRENT_AMPERE,
```
ELECTRIC_CURRENT_AMPERE

yayitazale reviewed (changes requested):
Due #53243 you must change these units of measurement

hesselonline committed:
Update const.py

Changed ELECTRICAL to ELECTRIC

hesselonline committed:
Merge branch 'dev' into wallbox-change-coordinator

yayitazale commented:
Now can be merged, please don't wait another 30 days

yayitazale commented:
@hesselonline maybe you need to check "I have reviewed two other open pull requests in this repository." to make this PR mergeable to the admins...

hesselonline commented:
@janiversen do you have time to give me a review on this PR? Thnx!

janiversen commented:
> @hesselonline maybe you need to check "I have reviewed two other open pull requests in this repository." to make this PR mergeable to the admins...

That is not true. It is not blocking anything, but of course it might secure the PR is reviewed faster.

janiversen commented:
@hesselonline you have a merge conflict. please do
git checkout dev
git pull
git checkout <your branch>
git rebase dev
git push -f

janiversen commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -5,41 +5,49 @@
 
-    def __init__(self, station, username, password, hass):
+    def __init__(
+        self, station, username: str, password: str, hass: HomeAssistant
```
Please add typing for all parameters.

janiversen commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -5,41 +5,49 @@
-        self._username = username
-        self._password = password
-        self._wallbox = Wallbox(self._username, self._password)
+        self._wallbox = Wallbox(username, password)
```
Why co you need line31 ? hass is automatically added as first parameters to all HA calls.

hesselonline replied:
If I leave it out (and add it as a parameter to class functions), I seem to (at least) be unable to access hass in the _async_update_data function. I get the following warning: wallbox integration not ready yet: _async_update_data() missing 1 required positional argument: 'hass'; Retrying in background.

Any tips?

janiversen replied:
Did a couple of tests, and seems the data_coordinator do not use hass. So your change is ok.

janiversen reviewed (changes requested):
I will review more later.

hesselonline committed:
Fix merge conflict caused by new energy dashboard functionality

hesselonline committed:
Added typing for class init params

added station typinh

janiversen commented on homeassistant/components/wallbox/number.py:
```diff
@@ -0,0 +1,58 @@
+    @property
+    def max_value(self):
+        """Return the maximum available current."""
+        return self.coordinator.data["max_available_power"]
```
```suggestion
        return self._coordinator.data["max_available_power"]
```

hesselonline replied:
I will fix this in an upcoming commit

janiversen commented on homeassistant/components/wallbox/number.py:
```diff
@@ -0,0 +1,58 @@
+    @property
+    def value(self):
+        """Return the state of the sensor."""
+        return self.coordinator.data["max_charging_current"]
```
```suggestion
        return self._coordinator.data["max_charging_current"]
```

janiversen commented on homeassistant/components/wallbox/number.py:
```diff
@@ -0,0 +1,58 @@
+        super().__init__(coordinator)
+        _properties = CONF_SENSOR_TYPES["max_charging_current"]
+        self._coordinator = coordinator
+        self._name = f"{config.title} {_properties[CONF_NAME]}"
```
```suggestion
        self._attr_name = f"{config.title} {_properties[CONF_NAME]}"
```

janiversen commented on homeassistant/components/wallbox/number.py:
```diff
@@ -0,0 +1,58 @@
+        self._name = f"{config.title} {_properties[CONF_NAME]}"
+        self._device_class = _properties[CONF_DEVICE_CLASS]
+
+    @property
```
Delete this (replaced by self._attr_name)

janiversen commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -34,10 +33,12 @@ class WallboxSensor(CoordinatorEntity, SensorEntity):
-        self._name = f"{config.title} {self._properties[CONF_NAME]}"
-        self._icon = self._properties[CONF_ICON]
-        self._unit = self._properties[CONF_UNIT_OF_MEASUREMENT]
+        _properties = CONF_SENSOR_TYPES[ent]
```
Do use _ for method variables, that is reserved for internal class variables.

Apart from that, it does not make the code more readable (I prefer to reference CONF_SENSOR_TYPES directly, in __init__), but that is a matter of taste.

hesselonline replied:
I took your suggestions and referred directly to the original dict.

janiversen commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -34,10 +33,12 @@ class WallboxSensor(CoordinatorEntity, SensorEntity):
-        self._unit = self._properties[CONF_UNIT_OF_MEASUREMENT]
+        _properties = CONF_SENSOR_TYPES[ent]
+
+        self._name = f"{config.title} {_properties[CONF_NAME]}"
```
```suggestion
        self._attr_name = f"{config.title} {_properties[CONF_NAME]}"
```
And remove the property name

janiversen commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -34,10 +33,12 @@ class WallboxSensor(CoordinatorEntity, SensorEntity):
+        _properties = CONF_SENSOR_TYPES[ent]
+
+        self._name = f"{config.title} {_properties[CONF_NAME]}"
+        self._icon = _properties[CONF_ICON]
```
```suggestion
        self._attr_icon = _properties[CONF_ICON]
```
and remove the property icon

janiversen commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -34,10 +33,12 @@ class WallboxSensor(CoordinatorEntity, SensorEntity):
+
+        self._name = f"{config.title} {_properties[CONF_NAME]}"
+        self._icon = _properties[CONF_ICON]
+        self._unit = _properties[CONF_UNIT_OF_MEASUREMENT]
```
```suggestion
        self._attr_unit_of_measurement = _properties[CONF_UNIT_OF_MEASUREMENT]
```
and remove the property unit_of_measurement

janiversen commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -34,10 +33,12 @@ class WallboxSensor(CoordinatorEntity, SensorEntity):
+        self._name = f"{config.title} {_properties[CONF_NAME]}"
+        self._icon = _properties[CONF_ICON]
+        self._unit = _properties[CONF_UNIT_OF_MEASUREMENT]
+        self._device_class = _properties[CONF_DEVICE_CLASS]
```
```suggestion
        self._attr_device_class = _properties[CONF_DEVICE_CLASS]
```
and remove the property device_class.

janiversen reviewed (changes requested):
Some review comments.

hesselonline committed:
Update homeassistant/components/wallbox/__init__.py

refer to DOMAIN in __init__

Co-authored-by: jan iversen <jancasacondor@gmail.com>

hesselonline commented on homeassistant/components/wallbox/number.py:
```diff
@@ -0,0 +1,58 @@
+
+from .const import CONF_CONNECTIONS, CONF_NAME, CONF_SENSOR_TYPES, DOMAIN
+
+CONF_STATION = "station"
```
That's a good point, honestly I don't even think those 2 consts are used anymore. I'm removing them from the number file

hesselonline committed:
Applied review suggestions in number file

hesselonline committed:
Merge branch 'wallbox-change-coordinator' of github.com:hesselonline/core into wallbox-change-coordinator

hesselonline committed:
Review comments for sensor file

hesselonline commented:
@janiversen any change that we can merge this one? That would allow me to continue with new functionality (support for energy dashboard and lock component)

frenck commented on homeassistant/components/wallbox/config_flow.py:
```diff
@@ -34,6 +34,31 @@ async def validate_input(hass: core.HomeAssistant, data):
+        self._station = None
+        self.reauth_schema = None
+
+    async def async_step_reauth(self, user_input=None):
+        """Perform reauth upon an API authentication error."""
+        self._reauth_entry = self.hass.config_entries.async_get_entry(
+            self.context["entry_id"]
+        )
+
+        self.reauth_schema = {
+            vol.Required(
+                CONF_STATION, default=self._reauth_entry.data[CONF_STATION]
+            ): str,
+            vol.Required(
+                CONF_USERNAME, default=self._reauth_entry.data[CONF_USERNAME]
+            ): str,
+            vol.Required(CONF_PASSWORD): str,
+        }
+
+        return await self.async_step_user()
```
This is a different feature. Please split different features into separate PRs.

Thanks 👍

hesselonline replied:
Yes, I might have been a bit overambitious, adding it all in one PR ;-). I'll move the reauth part to a seperate / next PR

janiversen commented:
> @janiversen any change that we can merge this one? That would allow me to continue with new functionality (support for energy dashboard and lock component)

I was just about to do a deeper dive, when I saw the comment from @frenck which I agree with.

hesselonline committed:
Removed the reauth flow from this PR, this change will be reintroduced in a seperate PR.

hesselonline committed:
Changed string and translations after removing reauth

hesselonline commented:
@frenck and @janiversen I removed the reauth part from this PR. I will add it in a seperate PR after this one has been merged

hesselonline committed:
Added more constants resolved test comments

hesselonline committed:
Changed minimum value of number to 6

hesselonline committed:
Moved static values to class attributes

hesselonline committed:
Removed typing and _hass for coordinator and renamed it

hesselonline committed:
Moved number component check to number.py, wallbox to setup_entry

hesselonline committed:
Fixed snakecase error

hesselonline committed:
Merge remote-tracking branch 'upstream/dev' into wallbox-change-coordinator

janiversen reviewed (changes requested):
I like your test coverage 👍

hesselonline commented on homeassistant/components/wallbox/number.py:
```diff
@@ -0,0 +1,50 @@
+    @property
+    def min_value(self):
+        """Return the maximum available current."""
+        return 0
```
I checked, and you are right -> its also 6 for me. I changed the value in the component to 6.

hesselonline commented:
@janiversen @frenck all the review comments are fixed, can we merge this PR now? Thnx!

janiversen reviewed (commented):
LGTM but I would like to hear the opinion of @frenck before approving.

ludeeus commented on homeassistant/components/wallbox/number.py:
```diff
@@ -0,0 +1,57 @@
+        self._device_class = _properties[CONF_DEVICE_CLASS]
+
+    @property
+    def min_value(self):
```
This is static and should be defined as a class attribute on the `WallboxNumber` class with `_attr_min_value = 6`

ludeeus commented on homeassistant/components/wallbox/number.py:
```diff
@@ -0,0 +1,57 @@
+        return self._coordinator.data[CONF_MAX_CHARGING_CURRENT_KEY]
+
+    @property
+    def device_class(self):
```
You already use `_attr_*` in this class, move this to the init as well with `self._attr_device_class = _properties[CONF_DEVICE_CLASS]`

ludeeus reviewed (dismissed):
This PR should really be 4 PR's

- Move to dataupdatecoordinator
- Use device class for sensor X,Y
- Add Number platform
- The restructure of sensor definitions, should move to `EntityDescription` (But do that in a separate PR, this is already crowded)

Here are a few things I found so far

hesselonline commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -5,37 +5,43 @@
-        self._password = password
-        self._wallbox = Wallbox(self._username, self._password)
+        self._wallbox = Wallbox(username, password)
         self._hass = hass
```
indeed, I did not know that self.hass existed. I changed it. I do not understand what you mean with the async_setup_entry comment. I don't think it would improve the code?

ludeeus replied:
You are constructing the `Wallbox` inside the init of the coordinator class, which means you need to pass authentication to it.
The  common way to do this is to construct the `Wallbox` object in `async_setup_entry` and then pass the object to the coordinator

hesselonline replied:
Done

hesselonline commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -5,37 +5,43 @@
 
-    def __init__(self, station, username, password, hass):
+    def __init__(
+        self, station: str, username: str, password: str, hass: HomeAssistant
```
removed typing

hesselonline commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -97,17 +125,31 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+            )
+        else:
+            try:
+                await wallbox.async_set_charging_current(
```
This check determines if number.py gets loaded, so moving it there doesn't seem right to me. It basically checks during setup if your account is authorized to alter the number value (via the api)

ludeeus replied:
While that might make sense now, it's more future-proof to move it.
What if you (or someone else) add other number entities to this integration that do not require this?

When that is added, this needs to move anyway, so it's better to do it now.

hesselonline replied:
OK, see your point now. Moved it to number.py

hesselonline commented:
> This PR should really be 4 PR's
> 
> * Move to dataupdatecoordinator
> * Use device class for sensor X,Y
> * Add Number platform
> * The restructure of sensor definitions, should move to `EntityDescription` (But do that in a separate PR, this is already crowded)
> 
> Here are a few things I found so far

For me this PR is about the number component, however as I was already working on the wallboxHub class changing it over to be a DataUpdateCoordinator was a small step to include. (also something that was requested after a previous PR). 
I added the device classes because it was suggested after a previous PR and it seemed like a small change, I don't know what you mean with the last restructure part. So basically, Coordinator and Number component are pretty intertwined for me, I can remove the device_classes from this PR if that helps?

ludeeus commented:
You can leave what you have in here now, but for your next PR try to keep it to changing/adding 1 thing.

hesselonline commented:
@ludeeus, @janiversen, @frenck any change on merging this one? It's stuck in the current state for a while now and I would like to continue with new features.

ludeeus commented:
The last commit was 13 days ago, there have been 2 rounds of reviews since

hesselonline commented:
hi @ludeeus. what do you mean with your remark? Is there an agreed upon waiting time between commits and reviews? If so, please let me know. In my experience I'm just trying to add some features to an integration, and in doing so I process the review comments given. Honestly, if I'm doing something wrong here let me know because its becoming quite a frustrating experience for me.

ludeeus commented:
What I mean is that this <https://github.com/home-assistant/core/pull/52786/commits/d71361827dc95273995ce380da2ac12068231017> is the last commit to the branch which is 13 days ago.
After that, both myself and Jan has requested changes.

hesselonline commented:
Yes, and I implemented those changes and asked for a reevaluation

ludeeus commented:
if you have implemented it, please push those changes up to this brnach.

hesselonline commented:
They have already been pushed to this branch. I do notice that the timestamp is off on them (commit is marked as 25-8, but it was actually a few days ago)

hesselonline commented:
I found the source for the wrong date, timesync was off on my dev VM. Apologies if that was the cause for some confusion.

hesselonline commented:
Yes, I understand, also something that you don't expect to be wrong. I only noticed it now because of the time difference between the comments and the code.

hesselonline committed:
Merge remote-tracking branch 'upstream/dev' into wallbox-change-coordinator

hesselonline commented:
Updated to include changes done in dev by PR 56308. All review comments have been dealt with.

hesselonline committed:
Merge remote-tracking branch 'upstream/dev' into wallbox-change-coordinator

hesselonline committed:
Merge remote-tracking branch 'upstream/dev' into wallbox-change-coordinator

hesselonline commented:
Hi, I don't know. Honestly, I stopped looking at this PR a month ago, I feel that I have done everything in my power to get it merged. I'll just wait and see what happens.

janiversen commented:
there is  famous movie by woody allan about this theme, The frase was “silent aggressive”. 

We are in hack october fest, with 376 PRs to review. I approved the PR, which enables it to be merged…..but I will say as you, lets wait and see what happens.

hesselonline committed:
merged httpstatus changes from dev
Merge branch 'dev' of github.com:home-assistant/core into wallbox-change-coordinator

bdraco reviewed (changes requested):
Please see comments above

hesselonline committed:
Add comment for number authorisation check

MartinHjelmare commented on homeassistant/components/wallbox/number.py:
```diff
@@ -0,0 +1,56 @@
+            coordinator.data[CONF_MAX_CHARGING_CURRENT_KEY]
+        )
+    except InvalidAuth:
+        pass
```
If we `return` here we can outdent below. Flat is better than nested. 👍

MartinHjelmare commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -6,37 +6,40 @@
+    """Wallbox Coordinator class."""
 
-    def __init__(self, station, username, password, hass):
+    def __init__(self, station, wallbox, hass):
```
Please put `hass` first in signatures that use it.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
